OpenTrons/opentrons-api
# feat(app, shared-data, step-generation): frontend low volume support

closes RAUT-714

# Overview

This adds low volume front-end support

# Test Plan

in PD/Step-generation:

sandbox: https://sandbox.designer.opentrons.com/app_pd-low-volume-support/

- create a Flex protocol with a p50 pipette (single or 8-channel) and a p1000 flex pipette (single or 8-channel). 
- Add 2x transfer steps (1 for each pipette)
- save and download the protocol and open up the json file. see that `configureForVolume` command occurs before the `aspirate` with the p50 pipette but not before the `aspirate` with the p1000 pipette.
- create a new Flex protocol with a p50 pipette (single or 8-channel). 
- this time add a mix step with mixing 1uL
- download the protocol, in the json file you should see only 1 `configureForVolume` command happening before the mix's `aspirate` and `dispense` commands

in the App:

- upload the protocol to the app and should pass protocol analysis. Check the run log command text for the `configureForVolume` commands.

# Changelog

in shared-data:
- add the types for `ConfigureForVolumeCreateCommand` and `ConfigureForVolumeRunTimeCommand`, add the optional `pushOut` param to `dispense` command

in app:
- add logic in `commandText` and add test cases

in PD:
- there are no visual changes to PD!

in Step-generation:
- create new `configureForVolume.ts` atomic command, add test for it
- add logic for emitting the command in `transfer.ts` before an `aspirate` happens and volume is 1uL and with the correct pipettes.
- add logic for emitting the command in `mix.ts` to happen before the set of mix `aspirate` and `dispense` commands.
- add a comment explaining that pushOut won't be specified in `dispense.ts`

# Review requests

see test plan

# Risk assessment

low
